src/components/data/test/DataTable-test.js

```diff
@@ -0,0 +1,299 @@
+      assert.calledWith(formatCell, rows[0], 'delicious');
+      assert.equal(formatCell.callCount, 3);
+    });
+  });
+
+  describe('interacting with row data', () => {
+    ['click', 'focus'].forEach(event => {
+      it(`invokes provided onSelectRow callback when a row is ${event}ed`, () => {
+        const onSelectRow = sinon.stub();
+        const wrapper = createComponent(DataTable, {
+          onSelectRow,
+        });
+
+        wrapper.find('tbody tr').first().simulate(event);
+
+        assert.calledWith(onSelectRow, fakeRows[0]);
+      });
+    });
+
+    it('invokes provided `onConfirmRow` callback when row is double-clicked', () => {
+      const onConfirmRow = sinon.stub();
+      const wrapper = createComponent(DataTable, {
+        onConfirmRow,
+      });
+
+      wrapper.find('tbody tr').first().simulate('dblclick');
+
+      assert.calledWith(onConfirmRow, fakeRows[0]);
+    });
+
+    it('invokes provided `onConfirmRow` callback when `Enter` is pressed on a row', () => {
+      const onConfirmRow = sinon.stub();
+      const wrapper = createComponent(DataTable, {
+        onConfirmRow,
+      });
+
+      wrapper.find('tbody tr').first().simulate('keydown', { key: 'Enter' });
+
+      assert.calledWith(onConfirmRow, fakeRows[0]);
+    });
+  });
+
+  context('when loading', () => {
+    it('renders a loading spinner', () => {
+      const wrapper = createComponent(DataTable, { loading: true });
+      assert.isTrue(wrapper.find('SpinnerSpokesIcon').exists());
+    });
+
+    it('does not render any data', () => {
+      const wrapper = createComponent(DataTable, { loading: true });
+      // One row, which holds the spinner
+      assert.equal(wrapper.find('tbody tr').length, 1);
+    });
+
+    it('still renders headings', () => {
+      const wrapper = createComponent(DataTable, { loading: true });
+      assert.equal(wrapper.find('thead tr th').length, 3);
+    });
+  });
+
+  context('when empty', () => {
+    it('shows an empty message if provided', () => {
+      const wrapper = createComponent(DataTable, {
+        emptyMessage: <strong>Nope</strong>,
+        rows: [],
+      });
+      assert.equal(wrapper.find('tbody tr td').at(0).text(), 'Nope');
+    });
+
+    it('still renders headings', () => {
+      const wrapper = createComponent(DataTable, {
+        emptyMessage: <strong>Nope</strong>,
+        rows: [],
+      });
+      assert.equal(wrapper.find('thead tr th').length, 3);
+    });
+  });
+
+  context('when in a Scroll context', () => {
+    let constrainedContainer;
+
+    function TestTableWithScroll() {
+      // Initially select the last data row
+      const [selectedRow, setSelectedRow] = useState(fakeRows[4]);
+      return (
+        <Scroll>
+          <DataTable
+            onSelectRow={setSelectedRow}
+            rows={fakeRows}
+            columns={fakeColumns}
+            selectedRow={selectedRow}
+            renderItem={(row, field) => <div>{row[field]}</div>}
```

Comment:
This test doesn't set an explicit height for rows that I can see. Does that mean the row height could vary depending on eg. the font size? If so it would probably be wise to set an explicit height here to make it easier to reason about what rows are visible at any time.

Reply:
I concur that the arithmetic ends up being fiddly and non-round. However, there's more (dimension-related) styling at play here than can be controlled from the `renderItem` callback. I can, for example, update it to:

renderItem={(row, field) => <div style="height:50px">{row[field]}</div>}

But this doesn't do what we might like it to — it results in a row height of `74px`. That's because there's other padding and whatnot in play here.

If we had a `renderRow` callback, we could precisely control row height. I don't think it's quite worth the tangent just now.

Reply:
Setting an explicit height in `renderItem` would at least produce a known _minimum_ height, and then a certain amount of padding could be assumed by the test with some tolerance for future minor style tweaks.